### ISSUE TITLE
False positive unresolved-symbol deftest test name due to strange interaction with :refer :all

### ISSUE THREAD

**Issue by tervorz** [borkdude/clj-kondo#1010]
**version**

clj-kondo v2020.09.09

**platform**

I'm using the native version on MacOS 10.15.6, however this is reproducible if I build master locally and run the jar against test files described in the problem section below.

**editor**

Not particularly applicable given this is reproducible outside it, but I'm using Emacs 26.3 (Doom) with flycheck.

**problem**

I get an unresolved-symbol error on the names of my tests when using `deftest`, but only when using `:refer :all` for `clojure.test` and also for any other second namespace. Very nuanced. So `(deftest my-test ...)` gives me an unresolved-symbol error on `my-test`.

**repro**

I wrote a unit test for this in my fork: https://github.com/tervorz/clj-kondo/blob/deftest-resolve-test-name-false-positive/test/clj_kondo/main_test.clj#L2134

Here is the test:
```
  (is (empty? (lint! (io/file "corpus" "deftest_resolve_test_name_fail.clj")
                     '{:linters {:unresolved-symbol {:level :error}
                                 :refer-all {:level :warning
                                             :exclude [clojure.test clojure.string]}}}))))
```

Contents of `deftest_resolve_test_name_fail.clj`:
```
(ns deftest-resolve-test-name-fail
  (:require [clojure.string :refer :all]
            [clojure.test :refer :all]))
(deftest my-test (is (blank? "")))
```

Linter output:
```
{:file "corpus/resolve_test_name_fail.clj",
 :row 4,
 :col 10,
:level :error,
:message "unresolved symbol my-test"}
```

This bug is pretty nuanced. It only surfaces if you have two `:refer :all` namespaces in your test file, and _only_ if `clojure.test :refer :all` comes after some other namespace's `:refer :all`. If `clojure.test` comes second, it works correctly, as can be seen in this unit test: https://github.com/tervorz/clj-kondo/blob/deftest-resolve-test-name-false-positive/test/clj_kondo/main_test.clj#L2130

**config**
```
{:linters {:unresolved-symbol {:level :error}
           :refer-all         {:level   :warning
                               :exclude [clojure.test clojure.string]}}}
```

**expected behavior**
No error.

I attempted to fix it on my own, but I couldn't pinpoint where the special handling was happening around deftest test names.

Thanks for your work on this! I'm new, but enjoying using it for my linting so far.

**Comment by borkdude** [comment]
@tervorz Thanks for posting the issue. Although it will attempt to resolve names to the right namespaces when possible. This seems like a case that should work. Feel free to dig deeper. Since clj-kondo doesn't encourage using `:refer :all`, I don't think this is high on the list, but I will keep the issue open.

**Comment by borkdude** [comment]
@tervorz I took another look at this issue. So what's happening: clj-kondo can resolve the vars to the right namespace (e.g. it can spot an incorrect arity pased to deftest in the above example when removing the test name and body), but for this is done in a later phase (in linters.clj).

So for refer-alls we might have to look into the (built-in) cache in the analyzer.clj phase and then do a similar thing as in your PR, but using the previously analyzed vars instead of a hard coded list.

**Comment by tervorz** [comment]
@borkdude Thanks for taking another look! Work got busy so I haven't revisited this since our last exchange, but I was still planning to circle back. I do have a working version in emacs on my machine now at least, but with the less than ideal solution from my PR.

Thanks for the tip. I have some time off in a week, so I'm hoping to revisit it then. But I saw you marked it for next release, which I appreciate. If you end up getting to it before me, I won't be offended 😄

**Comment by hugoduncan** [comment]
I've run into this in a project that defines a language embedded in Clojure.  The language is imported through several `:use` clauses, to reduce the verbosity of using the language.  It's a large code base and unlikely to change the style of imports.

An alternative solution might be to allow configuration of the fully resolved name of a symbol that appears unqualified, and is not otherwise defined.  This would be more work to use, but might be simpler to implement.

**Comment by borkdude** [comment]
@hugoduncan Thanks. That might be a possibility. Alternatively, how bad would it be to use explicit `:refer [foo bar]`  for your DSL?

**Comment by borkdude** [comment]
Maybe we could support something like this in the config for the `:use` and `:refer-all` linters:

```
{:use {:namespaces {foo.bar {:refer [a b c]}}}}
```
as a way to communicate to clj-kondo: when the namespace foo.bar is used (without `:only`), interpret that as `:require`+ `:refer [a b c]`?

**Comment by hugoduncan** [comment]
It would have to refer around 100 symbols, in many files, which I think would be brittle.

**Comment by hugoduncan** [comment]
The `:refer` in config would be workable.

**Comment by borkdude** [comment]
I've pondered about it some more. It would be most ergonomic for all if clj-kondo just read this data from the cache earlier. It will be a tad slower but this only when using `:use` or `:refer :all` which most namespace have not.

**Comment by borkdude** [comment]
@tervorz @hugoduncan Are you able to test commit https://github.com/borkdude/clj-kondo/commit/c20adcafb3d1a78d8ae79a32ffc1cb9f3d8398d9 from branch `early-read`? Binaries should be available in the builds, or test using the JVM. This commit reads from the cache in an earlier phase in case of a :refer :all or :use and thus should be able to recognize vars that were linted earlier or are built-in.

**Comment by tervorz** [comment]
@borkdude your fix works for my use case, thank you! (I built a binary off your branch and used within emacs, and the false positive vanished.)

**Comment by hugoduncan** [comment]
Definitely improves the number of symbols it finds.  I think the remaining ones unresolved are symbols that are imported from a different namespace (but that's a different issue).

**Comment by borkdude** [comment]
@hugoduncan Feel free to post a repro for the other issue.

**Comment by borkdude** [comment]
@hugoduncan Are you perhaps using `in-ns`?
